### ISSUE TITLE
Dumping and Restoring SCTP sockets with usrsctp

### ISSUE THREAD

**Issue by tA-bot-git** [sctplab/usrsctp#363]
Hi, 

I am running a program that requires SCTP connections to be dumped and restored. I would like to replace the kernel implementation with userland SCTP. I have built and installed usrsctplib and can run the example programs. To dump the SCTP sockets my program runs, would I just need to replace all the SCTP functions in my program with usr_sctp type functions, or is there an extra step I would need to do this? My apologies if this is a basic question, I am a novice at socket programming and would appreciate any help.

**Comment by tuexen** [comment]
It depends. Some functionality is identical. However, you can use `select()`/`poll()` in combination with user sockets.

So why do you need to dump and restore SCTP sockets?

**Comment by nxrighthere** [comment]
> So why do you need to dump and restore SCTP sockets?

@tuexen I guess the use-case is zero-downtime restarts of servers so that applications don't have to drop connections when restarting. This implemented in several QUIC-based libraries.

**Comment by tA-bot-git** [comment]
Thanks for the reply and suggestion, @tuexen. I'll have to look more into that, and get write back here if I have any more questions. 

As @nxrighthere rightly points out, I have a running application that needs to be dumped and restored almost immediately (zero-downtime), and the dumping tool I use doesn't support lksctp, so I'd ideally like to replace that with something else and that's what led me to look into usr-sctp.

**Comment by tA-bot-git** [comment]
Hi again, @tuexen. 

In my client code I have a sctp function such as: sctp_getpaddrs(). usrsctp has a similar implementation: _usrsctp_getsockopt()_. However, the socket type in the sctp implementation is an **integer**, while the one in usrsctp is a **struct socket**. Is it possible to fix this (turning the integer into a socket struct for usrsctp)?

**Comment by tuexen** [comment]
> Hi again, @tuexen.
> 
> In my client code I have a sctp function such as: sctp_getpaddrs(). usrsctp has a similar implementation: _usrsctp_getsockopt()_. However, the socket type in the sctp implementation is an **integer**, while the one in usrsctp is a **struct socket**. Is it possible to fix this (turning the integer into a socket struct for usrsctp)?

No. When using a kernel stack you are using an integer in userland which gets translated into a pointer to the socket in kernelland. Since we don't have this boundary here, the pointer type is exposed. This also makes sure that you don't mix SCTP userland sockets with normal kernel sockets...
